# **Project 1 - Sales Analysis - ETL**

## Objectives

Extract data from provided CSV files, cleaning it apply feature engineering if necessary then save to
a new CSV file


## Inputs

CSV files provided:

stores data-set.csv
sales data-set.csv
features data set.csv

Renamed files to a uniform naming convention:

Sales_Features_DataSet.csv
Sales_DataSet.csv
Sales_Stores_DataSet.csv

Note: original files are stored in Data/OriginalFiles


## Outputs

CSV file created from ETL etc:

Sales_CleanedDataForInsights.csv

## Additional Comments

Had this issue trying to use pip install - r requirements,txt to maunally installed:

numpy
pandas


Will install any more required libraries as discovered...

Error:

(.venv) rogerwilliams@Rogers-iMac Project1-SalesAnalysis % pip install -r requirements.txt
Collecting numpy==1.26.1 (from -r requirements.txt (line 1))
  Using cached numpy-1.26.1-cp312-cp312-macosx_10_9_x86_64.whl.metadata (61 kB)
Collecting pandas==2.1.1 (from -r requirements.txt (line 2))
  Using cached pandas-2.1.1-cp312-cp312-macosx_10_9_x86_64.whl.metadata (18 kB)
Collecting matplotlib==3.8.0 (from -r requirements.txt (line 3))
  Using cached matplotlib-3.8.0-cp312-cp312-macosx_10_12_x86_64.whl.metadata (5.8 kB)
Collecting seaborn==0.13.2 (from -r requirements.txt (line 4))
  Using cached seaborn-0.13.2-py3-none-any.whl.metadata (5.4 kB)
Collecting ydata-profiling==4.12.0 (from -r requirements.txt (line 5))
  Using cached ydata_profiling-4.12.0-py2.py3-none-any.whl.metadata (20 kB)
Collecting plotly==5.17.0 (from -r requirements.txt (line 6))
  Using cached plotly-5.17.0-py2.py3-none-any.whl.metadata (7.0 kB)
Collecting ppscore==1.1.0 (from -r requirements.txt (line 7))
  Using cached ppscore-1.1.0.tar.gz (46 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [20 lines of output]
      Traceback (most recent call last):
        File "/Users/rogerwilliams/Projects/Python/CourseProjects/Project1-SalesAnalysis/.venv/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/Users/rogerwilliams/Projects/Python/CourseProjects/Project1-SalesAnalysis/.venv/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/Users/rogerwilliams/Projects/Python/CourseProjects/Project1-SalesAnalysis/.venv/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 118, in get_requires_for_build_wheel
          return hook(config_settings)
                 ^^^^^^^^^^^^^^^^^^^^^
        File "/private/var/folders/wv/yty584vs1kq76d7zd3rv5wpw0000gn/T/pip-build-env-4o10ho_h/overlay/lib/python3.12/site-packages/setuptools/build_meta.py", line 333, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=[])
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/private/var/folders/wv/yty584vs1kq76d7zd3rv5wpw0000gn/T/pip-build-env-4o10ho_h/overlay/lib/python3.12/site-packages/setuptools/build_meta.py", line 301, in _get_build_requires
          self.run_setup()
        File "/private/var/folders/wv/yty584vs1kq76d7zd3rv5wpw0000gn/T/pip-build-env-4o10ho_h/overlay/lib/python3.12/site-packages/setuptools/build_meta.py", line 520, in run_setup
          super().run_setup(setup_script=setup_script)
        File "/private/var/folders/wv/yty584vs1kq76d7zd3rv5wpw0000gn/T/pip-build-env-4o10ho_h/overlay/lib/python3.12/site-packages/setuptools/build_meta.py", line 317, in run_setup
          exec(code, locals())
        File "<string>", line 12, in <module>
      ModuleNotFoundError: No module named 'pkg_resources'
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.

[notice] A new release of pip is available: 24.3.1 -> 26.1.2
[notice] To update, run: pip install --upgrade pip
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.


## Initalise Working Environment

In [1]:
#import libraries
import os
import numpy as np
import pandas as pd
import datetime

# Section 1 - Initalisation

## Intitialise all variables to be used in global stack 

In [2]:
#DataFrame vars for ETL
dfSales_DataSet = None
dfSales_DataSet_Work = None


#missing values check vars
intMissing = 0
intMissingPercent = 0
dfCheck = None

#duplicates check vars
intDuplicates = 0

#constant for project directory
CNST_STR_PROJECT_DIR = "Project1-SalesAnalysis"



## Set Current Directory To Base Project Directory

In [ ]:
#get project directory - default is jupyter notebook sub folder as that is where this file is located!
#so move back one to the project root path
# Source - https://stackoverflow.com/a/17726833
# Posted by chimpsarehungry
# Retrieved 2026-07-05, License - CC BY-SA 3.0

#get current folder
strTemp = os.getcwd()

#is the last part of the path the project directory?
if strTemp.endswith(CNST_STR_PROJECT_DIR):
   #if so do nothing 
   strCurrentDir = strTemp
else:    
   #get current working directory and move back one to the project root path
   strCurrentDir =  os.path.normpath(os.getcwd() + os.sep + os.pardir)
   #change directory
   os.chdir(strCurrentDir)

#confirm current directory is project directory
print(f"Current Directory: \n {os.getcwd()}")

Current Directory: 
 /Users/rogerwilliams/Projects/Python/CourseProjects/Project1-SalesAnalysis


# Section 1 - Extraction

- Read csv files
- Get schema info -> column and row numbers
- Get first 5 records
- Get list of column datatypes
- Get detailed schema infomation

## Read csv Files Into Variable For Processing

In [4]:
dfSales_DataSet = pd.read_csv(strCurrentDir + "/Data/Sales_DataSet.csv")


## Get Schema Info - Sales_DataSet

In [5]:
print(f"Schema Information \n Rows: {dfSales_DataSet.shape[0]} Columns: {dfSales_DataSet.shape[1]}")
dfSales_DataSet.head()

Schema Information 
 Rows: 421570 Columns: 5


,Store,Dept,Date,Weekly_Sales,IsHoliday
0,1,1,05/02/2010,24924.50,False
1,1,1,12/02/2010,46039.49,True
2,1,1,19/02/2010,41595.55,False
3,1,1,26/02/2010,19403.54,False
4,1,1,05/03/2010,21827.90,False


## Observations - Sales_DataSet
421,570 rows
5 columns

Columns:

Store, Dept, Date, Weekly_Sales, IsHoliday

In [6]:
#get column data types
print (f"Data Types: \n{dfSales_DataSet.dtypes}")

Data Types: 
Store             int64
Dept              int64
Date                str
Weekly_Sales    float64
IsHoliday          bool
dtype: object


## Observations - Sales_DataSet

Date is a string - needs conversion to datetime
Other column types look ok

In [6]:
#get schema infomation
print(f"More Information:\n {dfSales_DataSet.info()}")

<class 'pandas.DataFrame'>
RangeIndex: 421570 entries, 0 to 421569
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   Store         421570 non-null  int64  
 1   Dept          421570 non-null  int64  
 2   Date          421570 non-null  str    
 3   Weekly_Sales  421570 non-null  float64
 4   IsHoliday     421570 non-null  bool   
dtypes: bool(1), float64(1), int64(2), str(1)
memory usage: 13.3 MB
More Information:
 None


## Observations - Sales_DataSet

No missing values
File size: 13.3MB

---

# Section 2 - Transformation

- Get missing values percent
- Get duplicates (if any)


In [8]:
#get total missing values
intMissing = dfSales_DataSet.isnull().sum()

#calculate percent missing
intMissingPercent = ( intMissing / len(dfSales_DataSet) * 100)
dfCheck = pd.DataFrame( {"Missing Values": intMissing, "%" : intMissingPercent})

#show results
if dfCheck["Missing Values"].sum() !=0:
   print( dfCheck[dfCheck["Missing Values"] > 0])
else:
   print( "No Missing Values")     



No Missing Values


## Observations - Sales_DataSet

No missing values!

## Find Duplicates - Sales_DataSet

In [7]:
#get number of duplicates
intDuplicates = dfSales_DataSet.isnull().sum()
intDuplicates


Store           0
Dept            0
Date            0
Weekly_Sales    0
IsHoliday       0
dtype: int64

## Observations - Sales_DataSet

No duplicates!

## Get Unique Values - Sales_DataSet

In [8]:
#show unique values count
dfSales_DataSet.nunique()

Store               45
Dept                81
Date               143
Weekly_Sales    359464
IsHoliday            2
dtype: int64

## Observations - Sales_DataSet

Store has 45 unique values
Dept has 81 unique values
Date has 143 unique values
Weekly_Sales has 349,464 unique values
IsHoliday has 2 unique values

## Get Categorical Value Distribution

In [9]:
#get categorical distribution for object columns (non numeric)
for objColumn in ["Date"]:
    print( f"[{objColumn}] Value Distribution:")
    print( dfSales_DataSet[objColumn].value_counts())

[Date] Value Distribution:
Date
23/12/2011    3027
25/11/2011    3021
16/12/2011    3013
09/12/2011    3010
17/02/2012    3007
              ... 
09/07/2010    2903
16/07/2010    2901
20/08/2010    2901
27/08/2010    2898
13/08/2010    2896
Name: count, Length: 143, dtype: int64


## Observations - Sales_DataSet

Nothing obvious is visual as far as extreme deviations

In [10]:
## Create Copy Of DataFrame
dfSales_DataSet_Work = dfSales_DataSet.copy()

## Strip Spaces From Date Column

In [11]:
#remove spaces from date column

for objColumn in["Date"]:
    dfSales_DataSet_Work[objColumn] = dfSales_DataSet_Work[objColumn].str.strip()
    
#show first 5 rows to check    
dfSales_DataSet_Work.head()    

,Store,Dept,Date,Weekly_Sales,IsHoliday
0,1,1,05/02/2010,24924.50,False
1,1,1,12/02/2010,46039.49,True
2,1,1,19/02/2010,41595.55,False
3,1,1,26/02/2010,19403.54,False
4,1,1,05/03/2010,21827.90,False


## Observations - Sales_DataSet

No data corruptions

## Validate Numerical Values

see if Stores column has values less than 1
see if Dept column has values less than 1
see if Weekly_Sales column have values less then 0

In [12]:
#get count of how many values in Stores column are less than 1
intLessThanOne = (dfSales_DataSet_Work["Store"] < 1).sum()
#print result
intLessThanOne

np.int64(0)

---

## Observations - Sales_DataSet

Values for Store in acceptable limits

## Validate Dept Values

In [13]:
#get count of how many values in Dept column are less than 1
intLessThanOne = (dfSales_DataSet_Work["Dept"] < 1).sum()
#print result
intLessThanOne

np.int64(0)

## Observations - Sales_DataSet

Values for Store in acceptable limits

## Validate Weekly_Sales Values

In [14]:
#Get count of how many values in Weekly_Sales column are less than 1
intLessThanOne = (dfSales_DataSet_Work["Weekly_Sales"] < 1).sum()
#print result
intLessThanOne

np.int64(2006)

## Observations - Sales_DataSet

Values for Weekly_Values include negative values
This suggestss credits given, will leave data as is


## Data Transformations/Feature Engineering

Date - convert to datetime
Create column: IsCredit as boolean to allow easy sorting/grouping by negative/non negative values


In [15]:
#convert Date to datetime
dfSales_DataSet_Work["Date"] = dfSales_DataSet_Work["Date"].astype("datetime64[us]")
#show results
dfSales_DataSet_Work.dtypes

Store                    int64
Dept                     int64
Date            datetime64[us]
Weekly_Sales           float64
IsHoliday                 bool
dtype: object

In [16]:
#double check records
dfSales_DataSet_Work.head()

,Store,Dept,Date,Weekly_Sales,IsHoliday
0,1,1,2010-05-02,24924.50,False
1,1,1,2010-12-02,46039.49,True
2,1,1,2010-02-19,41595.55,False
3,1,1,2010-02-26,19403.54,False
4,1,1,2010-05-03,21827.90,False


## Observations - Sales_DataSet

Conversion successful

## Create Custom Column: IsCredit

In [17]:
#create scustom column IsCredit as boolean
dfSales_DataSet_Work["IsCredit"] = dfSales_DataSet_Work["Weekly_Sales"] < 0
#check output - filter for negative values
dfSales_DataSet_Work.query("Weekly_Sales <0")


,Store,Dept,Date,Weekly_Sales,IsHoliday,IsCredit
846,1,6,2012-10-08,-139.65,False,True
2384,1,18,2012-04-05,-1.27,False,True
6048,1,47,2010-02-19,-863.00,False,True
6049,1,47,2010-12-03,-698.00,False,True
6051,1,47,2010-08-10,-58.00,False,True
...,...,...,...,...,...,...
419597,45,80,2010-12-02,-0.43,True,True
419598,45,80,2010-02-19,-0.27,False,True
419603,45,80,2010-04-16,-1.61,False,True
419614,45,80,2010-02-07,-0.27,False,True


## Observations - Sales_DataSet

Column created sucessfully, data looks good

## Save Work DataFrame To CSV File

In [ ]:
#create list of columns new and old
lstColumns = [
                #original columns
                "Store","Dept","Date","Weekly_Sales","IsHoliday",
                #new colmns
                "IsCredit"
              ]

#create new DataFrame for cleaned data
dfClean = dfSales_DataSet_Work[lstColumns].copy()

#save dfSales_DataSet_Work
dfClean.to_csv(strCurrentDir + "/Data/CleanedDataSets/Sales_DataSet_Cleaned.csv")

# Conclusions and Next Steps